### PR TITLE
DS-597. Limit datasets with DATASETS_LIST and DATASETS_EXCLUDE

### PR DESCRIPTION
**Pull request by ngaranko**
# This Pull request contains changes to:

Router loading mechanics, allowing limitation of datasets loaded into memory using `DATASETS_LIST` and/or `DATASETS_EXCLUDE` environment variables.

Can be used to split APIs with sensitive information from generic APIs on infrastructure level.

# In case changes new features added

- [ ] Customer friendly documentation added into `docs/`.
- [x] Developer friendly documentation added into `DEVELOPMENT|README.md`

### CODE REVIEW

File: src/tests/test_dynamic_api/test_router.py [hunk 1]
```diff
@@ -29,3 +29,66 @@ def test_reload_delete(router, bommen_dataset):
 
     with pytest.raises(NoReverseMatch):
         assert reverse("dynamic_api:bommen-bommen-list")
+
+
+@pytest.mark.django_db
+def test_only_selected_datasets_loaded(
+    settings, router, bommen_dataset, gebieden_dataset, meldingen_dataset
+):
+    router.reload()
```
Comment (jjmurre) [inline src/tests/test_dynamic_api/test_router.py:38]: De `router.reload()` zou ook kunnen mbv. de `reloadrouter` fixture.

File: README.md [hunk 2]
```diff
@@ -197,3 +194,17 @@ OpenAPI schema can be verified on localhost using following command:
...
+
+# Limiting number of Datasets available via API
+
+In some cases it might be required to isolate datasets from others on infrastructure level, 
+for example sensitive information should not be published via Public APIs, but public information should be available in private API.
+
+This can be achieved by using `DATASETS_LIST` and `DATASETS_EXCLUDE` environment variables.
+
+In case API should expose only limited set of datasets - use `DATASETS_LIST` to load only provided datasets into memory.
+In case API should expose all datasets with exception to limited subset - use `DATASETS_EXCLUDE` to loxad all datasets except provided into memory.
```
Comment (lbam) [inline README.md:207]: "Loaded into memory" suggests datasets are resident in RAM, but they stay on disk, right?

Also, the syntax for these environment variables' values is missing. If I want to make foo and bar available, is that `foo bar`, `foo,bar`, `foo+bar`, ...

File: src/tests/test_dynamic_api/test_router.py [hunk 3]
```diff
@@ -29,3 +29,66 @@ def test_reload_delete(router, bommen_dataset):
...
+    assert reverse(
+        "dynamic_api:bommen-bommen-list"
+    )  # Bommen dataset reverse works, as dataset is registered
```
Comment (vdboor) [inline src/tests/test_dynamic_api/test_router.py:41]: misschien een comment boven de code plaatsen ivm black changes

File: README.md (hunk 2 above)
Comment (ngaranko) [reply README.md:207]: > "Loaded into memory" suggests datasets are resident in RAM, but they stay on disk, right?

Nope, they are loaded into memory for worker lifespan. Disk/Database usage during API construction is not advised, therefore everything is loaded into memory.

> Also, the syntax for these environment variables' values is missing. If I want to make foo and bar available, is that foo bar, foo,bar, foo+bar, ...
Good one, will add.

File: src/tests/test_dynamic_api/test_router.py (hunk 3 above)
Comment (ngaranko) [reply src/tests/test_dynamic_api/test_router.py:41]: Oh yes, forgot to check after black did black things.

File: README.md (hunk 2 above)
Comment (lbam) [reply README.md:207]: Ok, I hadn't figured that out yet.

Comment (jjmurre) [reply README.md:207]: Actually, they are loaded in memory. That is, the jsonschema definitions are read on startup of the Django application and converted into on-the-fly Django models in memory.

**Review by jjmurre — APPROVED** [review]
LGTM. We have to be sure that there is sensible feedback when Django does not start because of missing datasets.

**Comment by ngaranko** [comment]
@jjmurre  
> LGTM. We have to be sure that there is sensible feedback when Django does not start because of missing datasets.

We've got that thanks to @vdboor 

`DATASETS_LIST=meldingen ./manage.py runserver
Watching for file changes with StatReloader
2021-02-22 15:20:29,968 - django.utils.autoreload - INFO - Watching for file changes with StatReloader
Performing system checks...

Exception in thread django-main-thread:
Traceback (most recent call last):
    ...
    raise Exception(f"Dataset {dataset_id} is missing.")
Exception: Dataset gebieden is missing.
`

File: README.md (hunk 2 above)
Comment (lbam) [reply README.md]: This is confusing :) What is loaded into memory? The entire datasets, or just their schema definitions?

**Review by cvriel — APPROVED** [review]
Ziet er goed uit

File: README.md (hunk 2 above)
Comment (ngaranko) [reply README.md]: Whole API is built in memory and based on Dataset model, which is read only inside Router.

This PR is allowing users to limit datasets on which API is build. However it is easier to say "Datasets loaded into memory" as whole "in memory API" logic is complex and desires whole separate document on mechanics of API construction. (we have ticket for that).

Hope this clears things up, let me know if you want this to be added to current readme.

Comment (lbam) [reply README.md]: May I suggest an alternative phrasing:

* To expose only a subset of the datasets, use `DATASETS_LIST`.
* To expose all datasets with some exceptions, use `DATASETS_EXCLUDE`.

This avoids discussing implementation details, while also not suggesting that massive amounts of RAM need to be provisioned to get this service running.

File: README.md [hunk 4]
```diff
@@ -204,8 +204,8 @@ for example sensitive information should not be published via Public APIs, but p
 This can be achieved by using `DATASETS_LIST` and `DATASETS_EXCLUDE` environment variables.
 Both variables accept comma separated list of dataset ids e.g. `bommen,gebieden,meldingen` etc.
 
-In case API should expose only limited set of datasets - use `DATASETS_LIST` to load only provided datasets into memory.
-In case API should expose all datasets with exception to limited subset - use `DATASETS_EXCLUDE` to load all datasets except provided into memory.
+* To expose only a subset of the datasets, use `DATASETS_LIST`.
+* To expose all datasets with some exceptions, use `DATASETS_EXCLUDE`.
```
Comment (lbam) [inline README.md:208]: 👍

**Pull request merged**